syftr/storage.py

```
@@ -1230,3 +1233,118 @@ def iter_examples(self, partition="test") -> T.Iterator[QAPair]:
+    def check_paths_are_absolute(cls, v: Path) -> Path:
+        if not v.is_absolute():
+            raise ValueError(
+                f"The provided dataset path `{v}` must be an absolute path"
+            )
+        return v
+
+    @cached_property
+    def data_root(self) -> Path:
+        return Path(get_ray_temp_dir()) / "data" / "local_data"
+
+    def _get_qa_abspath(self) -> Path:
+        """
+        Ray worker-visible QA CSV path.
+        """
+        return self.qa_csv_path.absolute()
+
+    def _get_grounding_dir_abspath(self) -> Path:
+        """
+        Ray worker-visible grounding directory path.
+        """
+        return self.grounding_data_dir.absolute()
+
+    def _load_qa_dataset(self, partition: str = "test") -> datasets.Dataset:
+        """
+        This method reads a CSV file of QA pairs and returns all examples as the 'test' partition.
+        Other partitions ('sample', 'train', 'holdout') are not supported.
+        """
+        if partition != "test":
+            raise ValueError(
+                f"Only 'test' partition is supported in CustomLocalDataset (got '{partition}')"
+            )
+        dataset = pd.read_csv(self._get_qa_abspath())
+        return datasets.Dataset.from_pandas(dataset)
+
+    def _load_grounding_dataset(self) -> datasets.DatasetDict:
+        reader = SimpleDirectoryReader(input_dir=self._get_grounding_dir_abspath())
+        docs = reader.load_data()
+        dataset = datasets.Dataset.from_pandas(
+            pd.DataFrame(
+                [
+                    {
+                        "markdown": doc.text,
+                        "filename": doc.metadata.get("file_name", "unknown"),
+                    }
+                    for doc in docs
+                ]
+            )
+        )
+        return datasets.DatasetDict({"test": dataset})
+
+    def _row_to_qapair(self, row) -> QAPair:
+        """Dataset-specific conversion of row to QAPair struct."""
+        return QAPair(
```

use getter with defaults instead of defaults and try to populate all fields

---

explain optional fields in the example too

---

I don't see any attributes where a default is appropriate. It's better to KeyError when required data is missing than to silently ignore it. There are no optional fields either - anything outside of question, answer, and id is ignored.